tests/semantic_analyser.cpp

```diff
@@ -54,21 +55,24 @@ void test_for_warning(const std::string &input,
 void test(BPFtrace &bpftrace,
           bool mock_has_features,
           Driver &driver,
-          const std::string &input,
+          std::string_view input,
           int expected_result,
           std::string_view expected_error = {},
           bool safe_mode = true,
           bool has_child = false)
 {
+  if (!input.empty() && input[0] == '\n')
+    input.remove_prefix(1); // Remove initial '\n'
+
   std::stringstream out;
   std::stringstream msg;
   msg << "\nInput:\n" << input << "\n\nOutput:\n";
 
   bpftrace.safe_mode_ = safe_mode;
-  EXPECT_EQ(driver.parse_str(input), 0);
+  ASSERT_EQ(driver.parse_str(input), 0);
```

Comment:
Why do we want `ASSERT_EQ` here rather than EXPECT_EQ?

Reply:
In the semantic analyser tests, we use ASSERT for the earlier, non-semantic analysis, passes because we can't continue if they fail. We don't have any other error checking on them, so if we allow the test to continue after failing here then it will get a segmentation fault. (Try swapping back to EXPECT and then putting a syntax error in a semantic analysis test to confirm).

ASSERTs for early passes followed by EXPECTs for the pass we care about are the pattern that's used throughout the unit tests (see `test_for_warning()` just above this function, or `codegen/common.h`).

Reply:
thanks for detailing, that's the exact justification I was hoping for 👍